agent/submatview/store_test.go

```diff
@@ -0,0 +1,458 @@
+		defer cancel()
+
+		req2 = &fakeRequest{client: req.client, key: "key2", index: 22}
+
+		require.NoError(t, store.Notify(ctx, req2, cID, ch1))
+		go func() {
+			_, _ = store.Get(ctx, req2)
+		}()
+
+		// the original entry should still be at count 4
+		assertRequestCount(t, store, req, 4)
+		// the new entry should be at count 2
+		retry.Run(t, func(r *retry.R) {
+			assertRequestCount(r, store, req2, 2)
+		})
+	})
+
+	runStep(t, "end all the requests", func(t *testing.T) {
+		req.client.QueueEvents(
+			newEventServiceHealthRegister(10, 1, "srv1"),
+			newEventServiceHealthRegister(12, 2, "srv1"),
+			newEventServiceHealthRegister(13, 1, "srv2"),
+			newEventServiceHealthRegister(16, 3, "srv2"))
+
+		// The two Get requests should exit now that the index has been updated
+		retry.Run(t, func(r *retry.R) {
+			assertRequestCount(r, store, req, 2)
+		})
+
+		// Cancel the context so all requests terminate
+		cancel()
+		retry.Run(t, func(r *retry.R) {
+			assertRequestCount(r, store, req, 0)
+		})
+	})
+
+	runStep(t, "the expiry heap should contain two entries", func(t *testing.T) {
+		store.lock.Lock()
+		defer store.lock.Unlock()
+		e := store.byKey[makeEntryKey(req.Type(), req.CacheInfo())]
+		e2 := store.byKey[makeEntryKey(req2.Type(), req2.CacheInfo())]
+		require.Equal(t, 0, e2.expiry.Index())
+		require.Equal(t, 1, e.expiry.Index())
+
+		require.Equal(t, store.expiryHeap.Next().Entry, e2.expiry)
+	})
+}
+
+type testingT interface {
+	Helper()
+	Fatalf(string, ...interface{})
+}
+
+func assertRequestCount(t testingT, s *Store, req Request, expected int) {
+	t.Helper()
+
+	key := makeEntryKey(req.Type(), req.CacheInfo())
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	actual := s.byKey[key].requests
+	if actual != expected {
+		t.Fatalf("expected request count to be %d, got %d", expected, actual)
+	}
+}
+
+func TestStore_Run_ExpiresEntries(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	store := NewStore(hclog.New(nil))
+	ttl := 10 * time.Millisecond
+	store.idleTTL = ttl
+	go store.Run(ctx)
+
+	req := &fakeRequest{
+		client: NewTestStreamingClient(pbcommon.DefaultEnterpriseMeta.Namespace),
+	}
+	req.client.QueueEvents(newEndOfSnapshotEvent(2))
+
+	cID := "correlate"
+	ch1 := make(chan cache.UpdateEvent)
+
+	reqCtx, reqCancel := context.WithCancel(context.Background())
+	defer reqCancel()
+
+	require.NoError(t, store.Notify(reqCtx, req, cID, ch1))
+	assertRequestCount(t, store, req, 1)
+
+	// Get a copy of the entry so that we can check it was expired later
+	store.lock.Lock()
+	e := store.byKey[makeEntryKey(req.Type(), req.CacheInfo())]
+	store.lock.Unlock()
+
+	reqCancel()
+	retry.Run(t, func(r *retry.R) {
+		assertRequestCount(r, store, req, 0)
+	})
+
+	// wait for the entry to expire, with lots of buffer
+	time.Sleep(3 * ttl)
+
+	store.lock.Lock()
+	defer store.lock.Unlock()
+	require.Len(t, store.byKey, 0)
+	require.Equal(t, ttlcache.NotIndexed, e.expiry.Index())
+}
```

Comment:
I think these tests look great.

Possibly a TODO for later but what do you think about adding a slightly "fuzzy" test that starts up a bunch of concurrent `Notify` calls and asserts reasonable behaviour for each one. The value in the test isn't so much covering specific anomalies or cases but just verifying the concurrent behaviour works (i.e. all clients see the update), and when run with `-race` we have a better chance of catching any data race issues?

Reply:
Ya, that is an option. In my mind, if we were to do that, we should use a real backend not a fake one. It seems like a fake will never be close enough to the real behaviour that any such test wouldn't be meaningful with a fake.